election-server/src/db.rs

```diff
@@ -0,0 +1,310 @@
+        let get_ballot_submission = self
+            .client
+            .prepare_cached(
+                "SELECT transaction_hash, block_time, ballot, account, verified from ballots \
+                 WHERE transaction_hash = $1",
+            )
+            .await?;
+        let params: [&(dyn ToSql + Sync); 1] = [&transaction_hash.as_ref()];
+        let row = self
+            .client
+            .query_opt(&get_ballot_submission, &params)
+            .await?;
+        row.map(StoredBallotSubmission::try_from).transpose()
+    }
+
+    /// Get ballot submission by account address within the give range.
+    pub async fn get_ballot_submissions(
+        &self,
+        account_address: AccountAddress,
+        from: Option<usize>,
+        limit: usize,
+    ) -> DatabaseResult<Vec<StoredBallotSubmission>> {
+        let rows = if let Some(from) = from {
+            let get_ballot_submissions = self
+                .client
+                .prepare_cached(
+                    "SELECT id, transaction_hash, block_time, ballot, account, verified FROM \
+                     ballots WHERE account = $1 AND id < $2 ORDER BY id DESC LIMIT $3",
+                )
+                .await?;
+            let params: [&(dyn ToSql + Sync); 3] =
+                [&account_address.0.as_ref(), &(from as i64), &(limit as i64)];
+            self.client.query(&get_ballot_submissions, &params).await?
+        } else {
+            let get_ballot_submissions = self
+                .client
+                .prepare_cached(
+                    "SELECT id, transaction_hash, block_time, ballot, account, verified FROM \
+                     ballots WHERE account = $1 ORDER BY id DESC LIMIT $2",
+                )
+                .await?;
+            let params: [&(dyn ToSql + Sync); 2] = [&account_address.0.as_ref(), &(limit as i64)];
+            self.client.query(&get_ballot_submissions, &params).await?
+        };
+
+        rows.into_iter()
+            .map(StoredBallotSubmission::try_from)
+            .collect()
+    }
+}
+
+/// Wrapper around a database transaction
+pub struct Transaction<'a> {
+    /// The inner transaction
+    inner: &'a deadpool_postgres::Transaction<'a>,
+}
+
+impl<'a> From<&'a deadpool_postgres::Transaction<'a>> for Transaction<'a> {
+    fn from(inner: &'a deadpool_postgres::Transaction<'a>) -> Self { Self { inner } }
+}
+
+impl<'a> Transaction<'a> {
+    /// Set the latest height in the DB.
+    pub async fn set_latest_height(&self, height: AbsoluteBlockHeight) -> DatabaseResult<()> {
+        let set_latest_height = self
+            .inner
+            .prepare_cached("UPDATE settings SET latest_height = $1 WHERE id = true")
+            .await?;
+        let params: [&(dyn ToSql + Sync); 1] = [&(height.height as i64)];
+        self.inner.execute(&set_latest_height, &params).await?;
+        Ok(())
+    }
+
+    /// Insert a ballot submission into the DB.
+    pub async fn insert_ballot(
+        &self,
```

Comment:
Since these methods take `&self` already I don't see why the `Transaction` should wrap a reference to the inner transaction.